modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/snapshot/SnapshotRestoreProcess.java

```diff
@@ -0,0 +1,799 @@
+
+        if (meta != null && meta.consistentId().equals(cctx.localNode().consistentId().toString())
+            && meta.pageSize() != cctx.database().pageSize()) {
+            throw new IgniteCheckedException("Incompatible memory page size " +
+                "[snapshotPageSize=" + meta.pageSize() +
+                ", local=" + cctx.database().pageSize() +
+                ", snapshot=" + req.snapshotName() +
+                ", nodeId=" + cctx.localNodeId() + ']');
+        }
+
+        String pdsFolderName = ctx.pdsFolderResolver().resolveFolders().folderName();
+        List<File> cacheDirs = new ArrayList<>();
+        Map<String, StoredCacheData> cfgsByName = new HashMap<>();
+
+        // Collect cache configuration(s) and verify cache groups page size.
+        for (File snpCacheDir : cctx.snapshotMgr().snapshotCacheDirectories(req.snapshotName(), pdsFolderName)) {
+            String grpName = FilePageStoreManager.cacheGroupName(snpCacheDir);
+
+            if (!req.groups().contains(grpName))
+                continue;
+
+            ((FilePageStoreManager)cctx.pageStore()).readCacheConfigurations(snpCacheDir, cfgsByName);
+
+            File cacheDir = U.resolveWorkDirectory(ctx.config().getWorkDirectory(),
+                Paths.get(databaseRelativePath(pdsFolderName), snpCacheDir.getName()).toString(), false);
+
+            if (!cacheDir.exists())
+                cacheDir.mkdir();
+            else if (cacheDir.list().length > 0) {
+                throw new IgniteCheckedException("Unable to restore cache group, directory is not empty " +
+                    "[group=" + grpName + ", dir=" + cacheDir + ']');
+            }
+
+            cacheDirs.add(cacheDir);
+        }
+
+        Map<Integer, StoredCacheData> cfgsById = cfgsByName.isEmpty() ? Collections.emptyMap() :
+            cfgsByName.values().stream().collect(Collectors.toMap(v -> CU.cacheId(v.config().getName()), v -> v));
+
+        return new SnapshotRestoreContext(req.requestId(), req.snapshotName(), req.nodes(), cacheDirs, cfgsById);
+    }
+
+    /**
+     * Copy partition files and update binary metadata.
+     *
+     * @param opCtx Snapshot restore operation context.
+     * @param updateMeta Update binary metadata flag.
+     * @throws IgniteCheckedException If failed.
+     */
+    protected void restore(SnapshotRestoreContext opCtx, boolean updateMeta) throws IgniteCheckedException {
+        BooleanSupplier stopChecker = () -> opCtx.err.get() != null;
+        String pdsFolderName = ctx.pdsFolderResolver().resolveFolders().folderName();
+
+        if (updateMeta) {
+            File binDir = binaryWorkDir(
+                ctx.cache().context().snapshotMgr().snapshotLocalDir(opCtx.snpName).getAbsolutePath(), pdsFolderName);
+
+            if (stopChecker.getAsBoolean())
+                return;
+
+            // Check binary metadata compatibility.
+            ctx.cacheObjects().checkMetadata(binDir);
```

Comment:
Is it possible to unite the `checkMetadata` and `updateMetadata` methods? They are used both each time.

Reply:
Done